### PR TITLE
add example nb for hf transformers integration

### PR DESCRIPTION
**Pull request by kshitij12345**
# Description

Supersedes #116 

Add example notebook for HuggingFace Transformers Integration
---

## ❔ This change

- [x] adds a new feature
- [ ] fixes breaking code
- [ ] is cosmetic (refactoring/reformatting)

---

## ✔️ Pre-merge checklist

Mark which items have been completed:

- [x] Refactored code ([sourcery](https://sourcery.ai/))
- [x] Tested code locally
- [x] Added code to GitHub tests ([notebooks](workflows/test-notebooks.yml), [scripts](workflows/test-scripts.yml))
- [x] Updated GitHub [README](../README.md)

### 🪝 Run the following to both [pre-commit](https://pre-commit.com/) hooks and JupyterLab (turn on `format on save`)

- [x] Stripped outputs from notebooks - ([nbstripout](https://pypi.org/project/nbstripout/))
- [x] Sorted imports - ([isort](https://pycqa.github.io/isort/))
- [x] Formatted code - ([back](https://github.com/psf/black))

### 📄 Update documentation

Coordinate with @normandy7

- [ ] Updated content in the [docs](https://docs.neptune.ai)
- [ ] Added to all relevant docs index pages (e.g., [Examples](https://docs.neptune.ai/getting-started/examples), [Integrations intro](https://docs.neptune.ai/integrations-and-supported-tools/intro), [Model training index](https://docs.neptune.ai/integrations-and-supported-tools/model-training))
- [ ] Added to examples [README](../README.md)

---

## 🧪 Test Configuration

- OS:
- Python version: 3.9.12
- Neptune version: 0.16.7
- Affected libraries with version:
   * transformers==4.22.0 datasets==2.5.1 torch==1.12 scipy==1.6.3 scikit-learn==0.24.0 

---

## ✔️ Post-merge checklist

These tasks should be done after the code is merged to main. Create a subtask under the main ClickUp task to have these on your radar and share the subtask here.

__Post-merge subtask:__ `(ClickUp/JIRA task name)`

- [ ] Update relevant blogs with impacted code/functionality (Contact @Patrycja-J for the list of blogs)
- [ ] Add functionality under the Neptune column of the `competitor comparison sheet` (Coordinate with @SiddhantSadangi)
- [ ] Update integration README (Contact @Patrycja-J)

### CODE REVIEW

**Comment by kshitij12345** [comment]
@SiddhantSadangi Thanks for updating the branch with updated timeouts.

**Review by SiddhantSadangi — CHANGES_REQUESTED** [review]

File: integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+   "source": [
+    "# Using the 🤗 Transformers Integration\n",
+    "\n",
+    "Neptune provides an integration with 🤗 Transformers. So all you need to do to log metadata of the 🤗 Transformers training/finetuning is write a few lines of additional code and logging to Neptune is taken care of. \n",
+    "\n",
+    "You can integrate metadata tracking with Neptune either by:\n",
+    "* passing report_to = \"neptune\" to the Trainer arguments.\n",
+    "* setting up a Neptune callback and passing it to the Trainer callbacks.\n",
+    "\n",
+    "In this guide, we will look at the second option i.e. setting up a Neptune callback and passing it to the Trainer.\n",
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb:20]: Makes sense. Will add a note.

[the remark this answers is not in the document]
Comment (kshitij12345) [reply integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb:20]: I think the caveats should be mentioned in docs (rather than this example notebook/script).

cc: @normandy7

**Comment by SiddhantSadangi** [comment]
@kshitij12345 now that the scripts execute within ~3 mins, I think we can increase the epochs to 5

File: integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb (hunk 1 above)
[the remark this answers is not in the document]
Comment (normandy7) [reply integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb:20]: Added a note to the (new) docs.

File: integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+    "from getpass import getpass\n",
+    "my_api_token = getpass('Enter your Neptune API token: ')\n",
+    "\n",
+    "run = neptune.init_run(\n",
+    "    api_token=my_api_token,\n",
+    "    project=\"workspace-name/project-name\",  # replace with your own\n",
+    ")\n",
+    "```\n",
+    "\n",
+    "For example, if your workspace name is `ml-team` and the project name is `classification`, the project argument is: `project=\"ml-team/classification\"`\n",
```
[the remark this answers is not in the document]
Comment (normandy7) [reply integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb:123]: Yes, that seems right. It's from the notebook template. Unfortunately the standardization is lagging behind (which is OK in a way, as my first version of the template was not that good).

For now, let's use the template when creating new notebooks and revisit standardization once the new docs are up and running?

File: integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+   "metadata": {
+    "tags": [
+     "header",
+     "comment"
+    ]
+   },
+   "source": [
+    "# Using the 🤗 Transformers Integration\n",
+    "\n",
+    "Neptune provides an integration with 🤗 Transformers. So all you need to do to log metadata of the 🤗 Transformers training/finetuning is write a few lines of additional code and logging to Neptune is taken care of. \n",
```
Comment (normandy7) [inline integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb:14]:
```suggestion
    "Neptune provides an integration with 🤗 Transformers. All you need to do to log metadata of the 🤗 Transformers training/finetuning is add a few lines of additional code.\n",
```

File: integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+    "\n",
+    "You can integrate metadata tracking with Neptune either by:\n",
+    "* passing report_to = \"neptune\" to the Trainer arguments.\n",
```
Comment (normandy7) [inline integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb:17]:
```suggestion
    "* passing `report_to="neptune"` to the Trainer arguments.\n",
```

File: integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+    "* setting up a Neptune callback and passing it to the Trainer callbacks.\n",
+    "\n",
+    "In this guide, we will look at the second option i.e. setting up a Neptune callback and passing it to the Trainer.\n",
+    "\n",
+    "By the end of this guide, you will be able to use the HuggingFace Transformers integration to log\n",
```
Comment (normandy7) [inline integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb:22]:
```suggestion
    "By the end of this guide, you will be able to use the 🤗 Transformers integration to log:\n",
```

File: integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+    "* Train Loss\n",
```
Comment (normandy7) [inline integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb:23]:
```suggestion
    "* Train loss\n",
```

File: integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+    "* Evaluation Loss\n",
```
Comment (normandy7) [inline integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb:24]:
```suggestion
    "* Evaluation loss\n",
```

File: integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+    ]
+   },
+   "source": [
+    "## Start a run\n",
+    "\n",
+    "To connect your script to Neptune and create a new run, we tell Neptune:\n",
+    "* **Who you are** - with a Neptune API token\n",
+    "* **Where to send your data** - to a Neptune project\n",
+    "\n",
+    "The cell below lets you record data to the public project [common/quickstarts](https://app.neptune.ai/common/quickstarts) as an anonymous user."
```
Comment (normandy7) [inline integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb:77]: Should we open the Transformers example project to anonymous logging and change to that? @SiddhantSadangi

File: integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+    "    callbacks=[neptune_callback],\n",
+    "    tokenizer=tokenizer,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**NOTE**: One important thing to keep in mind is that `run` is stopped once `Trainer.train` is finished. Hence, we don't need to call `run.stop` explicitly (which is otherwise required when using Notebook)"
```
Comment (normandy7) [inline integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb:328]:
```suggestion
    "**Note**: The `run` object is stopped once `Trainer.train()` is finished. As such, we don't need to call `run.stop()` explicitly (which is otherwise required in interactive environments, such as Jupyter Notebook)."
```

File: integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+   "metadata": {},
+   "source": [
+    "#### Setting up the NeptuneCallback"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We pass the `run` that was created earlier to the NeptuneCallback. The NeptuneCallback will take care of logging the metadata during the training phase. You can customize the metadata that is logged by the callback by passing additional arguments to the Callback.\n",
```
Comment (normandy7) [inline integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb:287]:
```suggestion
    "We pass the `run` that was created earlier to the Neptune callback. The callback will take care of logging the metadata during the training phase. You can customize the metadata that is logged by passing additional arguments to the callback.\n",
```

File: integrations-and-supported-tools/transformers/scripts/Neptune_Transformers.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    AutoModelForSequenceClassification,
+    AutoTokenizer,
+    Trainer,
+    TrainingArguments,
+)
+from transformers.integrations import NeptuneCallback
+
+run = neptune.init_run(
+    api_token=neptune.ANONYMOUS_API_TOKEN,
+    project="common/quickstarts",
```
Comment (normandy7) [inline integrations-and-supported-tools/transformers/scripts/Neptune_Transformers.py:13]: Same question here: Should we enable anonymous logging for the Transformers project and use that here?

**Review by normandy7 — DISMISSED** [review]
Just wondering about the project we use for anonymous logging. Looks good otherwise.

File: integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (SiddhantSadangi) [reply integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb:123]: I think we should encourage the use of the [get my API token](https://app.neptune.ai/get_my_api_token) endpoint. It is built for this :)

File: integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb (hunk 8 above)
Comment (SiddhantSadangi) [reply integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb:77]: Nope. That project is used for our demos, and we don't want dirty runs there.

File: integrations-and-supported-tools/transformers/scripts/Neptune_Transformers.py (hunk 11 above)
Comment (SiddhantSadangi) [reply integrations-and-supported-tools/transformers/scripts/Neptune_Transformers.py:13]: Nope. That project is used for our demos, and we don't want dirty runs there.
However, the [huggingface-integration](https://app.neptune.ai/o/common/org/huggingface-integration/experiments?split=tbl&dash=charts&viewId=standard-view) project would be a better place to log runs than `quickstarts`

**Comment by SiddhantSadangi** [comment]
@kshitij12345 , @normandy7 
I think the [huggingface-integration](https://app.neptune.ai/o/common/org/huggingface-integration/experiments?split=tbl&dash=charts&viewId=standard-view) project would be a better place to log runs than `quickstarts`.
This is currently not open to anonymous logging and has some old runs by @twolodzko and @Raalsky , but we should be able to delete them and add `neptuner` to this project to anonymous logging.
Let me know what you think.

File: integrations-and-supported-tools/transformers/scripts/Neptune_Transformers.py (hunk 11 above)
Comment (kshitij12345) [reply integrations-and-supported-tools/transformers/scripts/Neptune_Transformers.py:13]: Agreed, @normandy7 did mention doing that. That reason I didn't do that was because I couldn't log with anonymous token. If that access can be provided easily then certainly we should point there.

**Comment by twolodzko** [comment]
@SiddhantSadangi fine for me.

File: README.md [hunk 12]
```diff
@@ -92,6 +92,7 @@ You can run every example with zero setup as an "ANONYMOUS" Neptune user (no reg
...
 | scikit-learn | [![docs]](https://docs.neptune.ai/integrations-and-supported-tools/model-training/sklearn) | [![neptune]](https://app.neptune.ai/o/common/org/sklearn-integration/e/SKLEAR-97/all?path=rfr_summary%2Fdiagnostics_charts&attribute=feature_importance) | [![github]](https://github.com/neptune-ai/examples/blob/main/integrations-and-supported-tools/sklearn/scripts) | [![colab]](https://colab.research.google.com/github/neptune-ai/examples/blob/master/integrations-and-supported-tools/sklearn/notebooks/Neptune_Scikit_learn.ipynb) |
 | skorch | [![docs]](https://docs.neptune.ai/integrations-and-supported-tools/model-training/skorch) | | | |
+| 🤗 Transformers | [![docs]](https://docs.neptune.ai/integrations-and-supported-tools/model-training/hugging-face) | [![neptune]](https://app.neptune.ai/o/showcase/org/project-text-summarization-hf/e/PROJ-138/dashboard/Custom-Dashboard-97370bc5-ee32-48ce-a630-fc806a370e13)  |  | [![colab]](https://colab.research.google.com/drive/1bJReZxX7fN_ZM74uSPKFuix5WlDRrKrv?usp=sharing) |
```
Comment (SiddhantSadangi) [inline README.md:95]: @kshitij12345 - this will need to be updated to point to notebook in github

**Review by SiddhantSadangi — CHANGES_REQUESTED** [review]

File: integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb (hunk 8 above)
Comment (kshitij12345) [reply integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb:77]: Updated.

File: README.md (hunk 12 above)
Comment (kshitij12345) [reply README.md:95]: Updated.

Comment (SiddhantSadangi) [reply README.md:95]: @kshitij12345 the notebook path should look like this
https://colab.research.google.com/github/neptune-ai/examples/blob/master/integrations-and-supported-tools/transformers/notebooks/Neptune_Transformers.ipynb
This ensures that colab always points to the latest notebook in git

File: integrations-and-supported-tools/transformers/scripts/run_examples.sh [hunk 13]
```diff
@@ -0,0 +1,9 @@
+echo "Installing requirements..."
+pip install -r requirements.txt
+
+echo "Setting Neptune credentials..."
+export NEPTUNE_API_TOKEN=ANONYMOUS
+export NEPTUNE_PROJECT=common/huggingface-integration
+
+echo "Running Neptune_Transformers.py..."
+python Neptune_Transformers.py
```
Comment (SiddhantSadangi) [inline integrations-and-supported-tools/transformers/scripts/run_examples.sh:9]: @kshitij12345 
`Neptune_Transformers_report_to.py` has to be added here as well

**Review by SiddhantSadangi — CHANGES_REQUESTED** [review]

File: README.md (hunk 12 above)
Comment (kshitij12345) [reply README.md]: Sorry. Forgot to push the change.

File: integrations-and-supported-tools/transformers/scripts/run_examples.sh (hunk 13 above)
Comment (kshitij12345) [reply integrations-and-supported-tools/transformers/scripts/run_examples.sh:9]: Ah. Missed that. Thanks!

**Comment by kshitij12345** [comment]
Except for Windows Python 3.8 failure, all other failures are expected.

Windows Python 3.8 failure looks flaky connection issue: https://github.com/neptune-ai/examples/actions/runs/3151258425/jobs/5124997271

**Comment by SiddhantSadangi** [comment]
> Windows Python 3.8 failure looks flaky connection issue: https://github.com/neptune-ai/examples/actions/runs/3151258425/jobs/5124997271

yeah, re-run succeeded

**Review by SiddhantSadangi — APPROVED** [review]

**Comment by SiddhantSadangi** [comment]
@normandy7 This is merged.
Can you update the checklist once the docs are updated?
Also, let's mention `Transformers` instead of `HuggingFace` in the index for consistency
